backend/src/lib/auth.ts

```diff
@@ -0,0 +1,142 @@
+  private siteRepository: Repository<Site>
+  private instrumentUploadRepository: Repository<InstrumentUpload>
+  private modelUploadRepository: Repository<ModelUpload>
+
+  constructor(conn: Connection) {
+    this.userAccountRepository = conn.getRepository<UserAccount>('user_account')
+    this.siteRepository = conn.getRepository<Site>('site')
+    this.instrumentUploadRepository = conn.getRepository<InstrumentUpload>('instrument_upload')
+    this.modelUploadRepository = conn.getRepository<ModelUpload>('model_upload')
+  }
+
+  uploadMiddleware = ({
+    permission,
+    isDataUpload = false,
+    isModelDataUpload = false,
+  }: UploadMiddlewareParams): RequestHandler => {
+    return async (req, res, next) => {
+      // Authenticator should handle authentication first and add username into res.locals
+      if (!res.locals.authenticated || !res.locals.username) {
+        console.error('Authorizator received unauthenticated request')
+        return next({ status: 401, errors: 'Unauthorized' })
+      }
+
+      let site: Site
+      if (!isModelDataUpload && !isDataUpload) {
+        if (
+          !Object.prototype.hasOwnProperty.call(req, 'body') ||
+          !Object.prototype.hasOwnProperty.call(req.body, 'site')
+        ) {
+          // Handle legacy upload
+          // username should match some site
+          const siteCandidate: Site | undefined = await this.siteRepository.findOne(res.locals.username)
+          if (siteCandidate === undefined) {
+            return next({ status: 400, errors: 'Add site to the request body' })
+          }
+          site = siteCandidate!
+        } else {
+          const siteCandidate: Site | undefined = await this.siteRepository.findOne(req.body.site)
+          if (siteCandidate === undefined) {
+            return next({ status: 422, errors: 'Site does not exist' })
+          }
+          site = siteCandidate!
+        }
+      } else {
+        if (isDataUpload) {
+          const uploadCandidate: InstrumentUpload | undefined = await this.instrumentUploadRepository.findOne(
+            { checksum: req.params.checksum },
+            { relations: ['site'] }
+          )
+          if (uploadCandidate === undefined) {
+            return next({ status: 422, errors: 'Checksum does not exist in the database' })
+          }
+          site = uploadCandidate!.site
+        } else if (isModelDataUpload) {
+          // modeldata upload checks site from db based on the checksum
+          const modelUploadCandidate: ModelUpload | undefined = await this.modelUploadRepository.findOne(
+            { checksum: req.params.checksum },
+            { relations: ['site'] }
+          )
+          if (modelUploadCandidate === undefined) {
+            return next({ status: 422, errors: 'Checksum does not exist in the database' })
+          }
+          site = modelUploadCandidate!.site
+        } else {
+          return next({ status: 500, errors: 'Oops, something went wrong' })
+        }
```

Comment:
Voisi tehdä geneerisesti:

```suggestion
        const repo: Repository<Upload> = isDataUpload ? this.instrumentUploadRepository : this.modelUploadRepository
        const uploadCandidate = await repo.findOne(
          { checksum: req.params.checksum },
          { relations: ['site'] }
        )
        if (uploadCandidate === undefined) {
          return next({ status: 422, errors: 'Checksum does not exist in the database' })
        }
        site = uploadCandidate!.site
```